bundles/automation/org.eclipse.smarthome.automation.module.script.rulesupport/src/main/java/org/eclipse/smarthome/automation/module/script/rulesupport/shared/simple/SimpleRule.java

```diff
@@ -12,18 +12,306 @@
+        return tags;
+    }
+
+    /**
+     * This method is used to specify the {@link Rule}'s assigned tags.
+     *
+     * @param ruleTags the {@link Rule}'s assigned tags.
+     */
+    @SuppressWarnings("null")
+    public void setTags(Set<String> ruleTags) {
+        tags = ruleTags != null ? ruleTags : new HashSet<>();
+    }
+
+    @Override
+    @Nullable
+    public String getDescription() {
+        return description;
+    }
+
+    /**
+     * This method is used to specify human-readable description of the purpose and consequences of the
+     * {@link Rule}'s execution.
+     *
+     * @param ruleDescription the {@link Rule}'s human-readable description, or {@code null}.
+     */
+    public void setDescription(@Nullable String ruleDescription) {
+        description = ruleDescription;
+    }
+
+    @Override
+    public Visibility getVisibility() {
+        return visibility;
+    }
+
+    /**
+     * This method is used to specify the {@link Rule}'s {@link Visibility}.
+     *
+     * @param visibility the {@link Rule}'s {@link Visibility} value.
+     */
+    @SuppressWarnings("null")
+    public void setVisibility(Visibility visibility) {
+        this.visibility = visibility == null ? Visibility.VISIBLE : visibility;
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.eclipse.smarthome.automation.Rule#getConfiguration()
+     */
+    @Override
+    public Configuration getConfiguration() {
+        return configuration;
+    }
+
+    /**
+     * This method is used to specify the {@link Rule}'s {@link Configuration}.
+     *
+     * @param ruleConfiguration the new configuration values.
+     */
+    @SuppressWarnings("null")
+    public void setConfiguration(Configuration ruleConfiguration) {
+        this.configuration = ruleConfiguration == null ? new Configuration() : ruleConfiguration;
+    }
+
+    @Override
+    public List<ConfigDescriptionParameter> getConfigurationDescriptions() {
+        return configDescriptions;
+    }
+
+    /**
+     * This method is used to describe with {@link ConfigDescriptionParameter}s
+     * the meta info for configuration properties of the {@link Rule}.
+     */
+    @SuppressWarnings("null")
+    public void setConfigurationDescriptions(List<ConfigDescriptionParameter> configDescriptions) {
+        this.configDescriptions = configDescriptions == null ? new ArrayList<>() : configDescriptions;
+    }
+
+    @Override
+    public List<Condition> getConditions() {
+        return conditions;
+    }
+
+    /**
+     * This method is used to specify the conditions participating in {@link Rule}.
+     *
+     * @param conditions a list with the conditions that should belong to this {@link Rule}.
+     */
+    @SuppressWarnings("null")
+    public void setConditions(List<Condition> conditions) {
+        this.conditions = conditions == null ? new ArrayList<>() : conditions;
+    }
+
+    @Override
+    public List<Action> getActions() {
+        return actions;
```

Comment:
We should return a copy, not the internal collection that could be modified by the caller.

Reply:
The solution is to use immutable collections. How it is accomplished is an implementation detail.
In this case I would prefer to create unmodifiable collection.
I will update the pull request.

Reply:
You should consider to hold the unmodifiable collection view and return that view but not return a unmodifiable view of the internal modifiable collection. This could break the consumer.

Reply:
If someone (like our tests) uses the returned list to update the Rule instance, we'll still break his code, but this time silently. I would prefer to break it with an exception.
Also one of the reasons for this change was to avoid making internal copies.

Reply:
Note that you are discussing here on the SimpleRule object for the jsr223 support - I am not sure, whether it was a feature here that you can actually modify stuff; I at least tried to keep jsr223 out of scope of this PR as much as possible I didn't want to break its APIs. So unless you have any strong reason to believe that something much be addressed here right away, I would suggest to leave those classes as they are and work on it in a subsequent PR, possibly with the involvement of @smerschjohann or @lewie.